Making list_trees work recursively through directories

KonstantinSchubert opened the pull request:
Many people put their TTrees into TDirectories .
The `root_numpy.list_trees` function does not find list TTrees, because it searches only in the top-level of the ROOT file.

This pull request implements a recursive search through the directory tree. `root_numpy.list_trees` now returns the full paths to all Trees:
```
In [3]: root_numpy.list_trees(file_path)                                                                                                   
Out[3]: ['Bu2LLK_meLine/DecayTree', 'GetIntegratedLuminosity/LumiTuple']
````

Please go easy on me, this is the first time I'm using Cython.

Some open questions from my side:

1. If backwards compatibility is an issue, maybe the function signature should be changed to:
`root_numpy.list_trees(file_path, recursive=False) `

2. Maybe it would make sense to analogously change the behavior of `list_directories`

3. Because I am trying to understand Cython:  Why can I assign to `next_dir` in (line 75)[https://github.com/rootpy/root_numpy/compare/master...KonstantinSchubert:master#diff-06942ee9ba64664fb84fa1b8cd8ce667R75] without cdef-ing this variable before? I don't see that anywhere else in your code.

KonstantinSchubert commented:
One more thing: My compiler/Cython combination seems to have a problem with line 85 in converters.pyx.  I had to comment stub out that function, which means that the "*.cpp" files are missing that function too. 

They shouldn't be published like that. 

When/if you accept the pull request, you could maybe check out, compile and push the correctly compiled "*.cpp" files immediately?

KonstantinSchubert commented:
Regarding the broken travis-ci, maybe switching to anaconda is an option? https://github.com/ibab/root_pandas/commit/6b44dc367a7b20864b676de32e8795f51677d468

ndawe commented:
Rebasing on master will fix the failing tests. root_numpy is now using conda thanks to @remenska 

> Because I am trying to understand Cython: Why can I assign to next_dir in (line 75)[master...KonstantinSchubert:master#diff-06942ee9ba64664fb84fa1b8cd8ce667R75] without cdef-ing this variable before? I don't see that anywhere else in your code.

I think cython can infer the type or sometimes it will fall back on a PyObject wrapper. You can see what cython is doing here by looking at the annotated lines in the generated .cpp

ndawe commented:
This PR will supersede #232

ndawe commented:
Could you instead implement a list_objects (and list_objects_recursive) function to handle any class name (or list of class names ie TTree TNtuple). Then list_trees and list_directories can call this central function.

KonstantinSchubert commented:
Okay, I will do that. 

* I think it is safer and more in line with the scope of the library if list_recursive to list only TDirectories, TTrees and TNtuples for now.

* I think  list_objects will return a list of tuples [(path, type-as-string), ...] or a dict {path: type-as-string, ...}
I'll see what is easier to handle.

I hope these points are okay.

Do you prefer rebase on master over a merge? I assume so from you comment.

KonstantinSchubert commented:
Is it possible that the tests do not re-compile the .cpp files?

KonstantinSchubert commented:
I hope to have addressed your comments.

Please be warned that I had to comment out the following lines in converters.pyx before compiling:
```
for i in range(numele):
      tmp[i] = src.at(i)
```

This hack is compiled into `root_numpy/src/_librootnumpy.cpp`
The tests don't catch it, but this needs to be re-compiled with the two lines added back in.

I will try to fix my compiler error and do it. Maybe you have an idea what could be causing the error?

```
root_numpy/src/converters.pyx: In function ‘int __pyx_f_13_librootnumpy_create_numpyarray_vectorbool(void*, std::vector<bool>*)’:
root_numpy/src/converters.pyx:85:17: error: no match for ‘operator=’ (operand types are ‘__Pyx_FakeReference<bool>’ and ‘std::vector<bool>::reference {aka std::_Bit_reference}’)
         tmp[i] = src.at(i)
                 ^
root_numpy/src/converters.pyx:85:17: note: candidates are:
root_numpy/src/_librootnumpy.cpp:224:7: note: __Pyx_FakeReference<bool>& __Pyx_FakeReference<bool>::operator=(const __Pyx_FakeReference<bool>&)
 class __Pyx_FakeReference {
       ^
root_numpy/src/_librootnumpy.cpp:224:7: note:   no known conversion for argument 1 from ‘std::vector<bool>::reference {aka std::_Bit_reference}’ to ‘const __Pyx_FakeReference<bool>&’
root_numpy/src/_librootnumpy.cpp:224:7: note: __Pyx_FakeReference<bool>& __Pyx_FakeReference<bool>::operator=(__Pyx_FakeReference<bool>&&)
root_numpy/src/_librootnumpy.cpp:224:7: note:   no known conversion for argument 1 from ‘std::vector<bool>::reference {aka std::_Bit_reference}’ to ‘__Pyx_FakeReference<bool>&&’
error: command 'gcc' failed with exit status 1
make: *** [inplace] Error 1
```

ndawe commented:
Just quickly about that cython error: this seems to be an issue with versions of cython newer than 0.21.2. I need to follow that up with the Cython devs. It will compile with cython 0.21.2.

ndawe commented:
Yes, rebase instead of merge.

ndawe commented:
Just remove #34075e1 and then rebase.

ndawe commented:
I fixed the cython error in #244 so you can use the latest cython without issues.

KonstantinSchubert commented:
Thanks. Done. Should I also update the documentation?

ndawe commented:
Nice. Reviewing now

ndawe commented on root_numpy/src/tree.pyx:
```diff
@@ -2,43 +2,22 @@ include "converters.pyx"
+    objects = list_objects(fname)
+    return [key for key, val in objects.items() if val == 'TDirectoryFile']
+
+def list_objects(fname):
```
I think list_objects should also take an optional list of class names and only populate the dictionary with keys that match those names. Then in list_trees above you could have ``objects = list_objects(fname, types=['TTree', 'TNtuple'])``.

ndawe commented on root_numpy/src/tree.pyx:
```diff
@@ -2,43 +2,22 @@ include "converters.pyx"
-            ret.append(key.GetName())
-    rfile.Close()
+        raise IOError("cannot read {0}".format(fname))
+    dictionary = {}
```
Call this ``objects``

KonstantinSchubert commented:
I tried to reflect your line comments.

ndawe commented on root_numpy/src/tree.pyx:
```diff
@@ -2,43 +2,19 @@ include "converters.pyx"
-    del rfile
-    return list(ret.keys())
-
+    return list_objects(fname, types=['TTree', 'TNtuple']).keys()
```
To get the python 3 tests passing you need a list(mydict.keys()) here.

ndawe commented on root_numpy/src/tree.pyx:
```diff
@@ -2,43 +2,19 @@ include "converters.pyx"
 
 def list_directories(fname):
-    cdef TFile* rfile = Open(fname, 'read')
+    return list_objects(fname, types=['TDirectoryFile']).keys()
```
Same here: `return list(list_objects(fname, types=['TDirectoryFile']).keys())`

KonstantinSchubert commented:
I learned a thing or two about `git rebase today ...

And sorry for not paying attention to the tests.

ndawe commented:
No problem. This also got me a few times when using python 3 (along with many other things...)

Thanks a lot for this pull request. I'm clicking the green button. :beers:
